### PR TITLE
add  gpt2 model

### PR DESCRIPTION
**Pull request by ZHUI**

### CODE REVIEW

File: PaddleNLP/examples/language_model/gpt2/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+3. 环境依赖
+
+    该模型使用PaddlePaddle，关于环境依赖部分，请先参考PaddlePaddle[安装说明](https://www.paddlepaddle.org.cn/documentation/docs/zh/install/index_cn.html)关于环境依赖部分的内容。
+
+
+### 数据准备
+
+#### Pre-training数据准备
+
+预训练公开数据集: [OpenWebTextCorpus](https://skylion007.github.io/OpenWebTextCorpus/)。此数据集包含了大约37GB的网页文本内容。为了方便用户运行测试本模型，本项目提供了处理好的300M的训练样本：
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/README.md]: 这里提示一下只是提供数据来源的具体，而且这里只是简单了处理，还有一个具体的数据逻辑，要加上具体github链接

File: PaddleNLP/examples/language_model/gpt2/README.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+```shell
+mkdir data && cd data
+wget https://paddlenlp.bj.bcebos.com/models/transformers/gpt2/train.data.json_ids.npz
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/README.md:67]: 这里需要加上process_data.py的处理逻辑

File: PaddleNLP/examples/language_model/gpt2/README.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+- `batch_size` 训练的batch大小
+- `device` 训练设备
+
+### 单机多卡
+
+同样，可以执行如下命令实现八卡训练：
+
+```shell
+export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
+python -m paddle.distributed.launch --gpus "0,1,2,3,4,5,6,7" run_pretrain.py --model_name_or_path gpt2-small-en \
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/README.md:112]: 这里不能export CUDA_VISIBLE_DEVICES, 都是unset一下

File: PaddleNLP/examples/language_model/gpt2/README.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    --output_dir "output"\
+    --weight_decay 0.01\
+    --grad_clip 1.0\
+    --max_steps 500000\
+    --save_steps 100000\
+    --warmup_rate 0.01\
+    --batch_size 8\
+    --device gpu
+
+```
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/README.md:123]: 这里可能还要加一个CPM的演示逻辑，待确认

File: PaddleNLP/paddlenlp/transformers/tokenizer_utils.py [hunk 5]
```diff
@@ -306,8 +306,8 @@ def save_resources(self, save_directory):
         Args:
             save_directory (str): Directory to save files into.
         """
-        assert hasattr(self, 'vocab') and len(
-            self.resource_files_names) == 1, "Must overwrite `save_resources`"
+        #assert hasattr(self, 'vocab') and len(
+        #    self.resource_files_names) == 1, "Must overwrite `save_resources`"
```
Comment (wawltor) [inline PaddleNLP/paddlenlp/transformers/tokenizer_utils.py]: 这块去掉注释

File: PaddleNLP/examples/language_model/gpt2/data.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
+import os
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/data.py:16]: 这里加一个copyright，同时把Megatron的copyright加上去, 同时时间是2021 ，其它的copyright也改成2021年

File: PaddleNLP/examples/language_model/gpt2/data.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+
+def _build_doc_idx(documents, num_epochs, np_rng, separate_last_epoch):
+    """Build an array with length = number-of-epochs * number-of-dcuments.
+    Each index is mapped to a corresponding document."""
+    if not separate_last_epoch or num_epochs == 1:
+        doc_idx = np.mgrid[0:num_epochs, 0:len(documents)][1]
+        doc_idx[:] = documents
+        doc_idx = doc_idx.reshape(-1)
+        doc_idx = doc_idx.astype(np.int32)
+        #np_rng.shuffle(doc_idx)
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/data.py]: 打开

File: PaddleNLP/examples/language_model/gpt2/data.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+
+def _build_shuffle_idx(num_samples, total_size, np_rng):
+    dtype_ = np.uint32
+    if total_size >= (np.iinfo(np.uint32).max - 1):
+        dtype_ = np.int64
+
+    shuffle_idx_first = np.arange(
+        start=0, stop=num_samples, step=1, dtype=dtype_)
+    #np_rng.shuffle(shuffle_idx_first)
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/data.py]: 打开

File: PaddleNLP/examples/language_model/gpt2/data.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    if num_samples == total_size:
+        return shuffle_idx_first
+
+    shuffle_idx_last = np.arange(
+        start=num_samples, stop=total_size, step=1, dtype=dtype_)
+    #np_rng.shuffle(shuffle_idx_last)
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/data.py]: 打开

File: PaddleNLP/examples/language_model/gpt2/lr.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+class CosineAnnealingWithWarmupDecay(LRScheduler):
+    def __init__(self,
+                 max_lr,
+                 min_lr,
+                 warmup_step,
+                 decay_steps,
+                 last_epoch=-1,
+                 verbose=False):
+
+        self.decay_steps = decay_steps
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/lr.py]: 这里的decay_steps改成decay_step吧

File: PaddleNLP/examples/language_model/gpt2/lr.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+                                                             verbose)
+
+    def get_lr(self):
+        if self.warmup_step > 0 and self.last_epoch <= self.warmup_step:
+            return float(self.max_lr) * (self.last_epoch) / self.warmup_step
+
+        if self.last_epoch > self.decay_steps:
+            return self.min_lr
+
+        num_steps_ = self.last_epoch - self.warmup_step
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/lr.py]: num_steps_改成num_step

File: PaddleNLP/examples/language_model/gpt2/process_data.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import argparse
+import json
+import multiprocessing
+import pickle
+import mmap
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/process_data.py]: 无效import

File: PaddleNLP/examples/language_model/gpt2/process_data.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import argparse
+import json
+import multiprocessing
+import pickle
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/process_data.py]: 无效import

File: PaddleNLP/examples/language_model/gpt2/run.sh [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+    --max_lr 0.00015\
+    --min_lr 0.00001\
+    --weight_decay 0.01\
+    --grad_clip 1.0\
+    --max_steps 500000\
+    --save_steps 100000\
+    --decay_steps 320000\
+    --warmup_rate 0.01\
+    --batch_size 8\
+    --device gpu
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/run.sh]: 这里建个scripts的目录吧，把目前的单机单卡版本和单机多卡的版本的脚本搞上去

File: PaddleNLP/examples/language_model/gpt2/run_pretrain.py [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+
+    global_step = 0
+    tic_train = time.time()
+    for epoch in range(args.num_train_epochs):
+        files = [
+            os.path.join(args.input_dir, f) for f in os.listdir(args.input_dir)
+            if (os.path.isfile(os.path.join(args.input_dir, f)) and "npz_"
+                not in str(f))
+        ]
+        #files.sort()
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/run_pretrain.py]: 打开

File: PaddleNLP/examples/language_model/gpt2/run_pretrain.py [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+def create_pretrained_dataset(args, input_path, worker_init, worker_index,
+                              eod_id):
+    train_data = GPT2Dataset(
+        file_path=input_path,
+        worker_index=worker_index,
+        num_samples=args.batch_size * args.max_steps,
+        eod_id=eod_id,
+        seed=args.seed + worker_index)
+    train_batch_sampler = paddle.io.BatchSampler(
+        train_data, batch_size=args.batch_size, shuffle=False)
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/run_pretrain.py]: shuffle打开

File: PaddleNLP/examples/language_model/gpt2/run_pretrain.py [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+        idx = 0
+    else:
+        idx = paddle.distributed.get_rank()
+    random.seed(args.seed + idx)
+    np.random.seed(args.seed + idx)
+    paddle.seed(args.seed + idx)
+
+
+def do_train(args):
+    assert args.device in ["cpu", "gpu", "xpu"], "invalid device!"
```
Comment (wawltor) [inline PaddleNLP/examples/language_model/gpt2/run_pretrain.py]: 这块assert稍微正式一点，提示信息长一点

File: PaddleNLP/paddlenlp/transformers/gpt2/modeling.py [hunk 18]
```diff
@@ -0,0 +1,611 @@
...
+        key = query if key is None else key
+        value = query if value is None else value
+        # compute q ,k ,v
+        if use_cache is False:
+            q, k, v = self._prepare_qkv(query, key, value, use_cache, cache)
+        else:
+            q, k, v, cache = self._prepare_qkv(query, key, value, use_cache,
+                                               cache)
+        # scale dot product attention
+        # TODO(guosheng): use tensor.matmul, however it doesn't support `alpha`
```
Comment (wawltor) [inline PaddleNLP/paddlenlp/transformers/gpt2/modeling.py]: 这块的TODO删除掉

File: PaddleNLP/paddlenlp/transformers/gpt2/modeling.py [hunk 19]
```diff
@@ -0,0 +1,611 @@
...
+        product = layers.matmul(
+            x=q, y=k, transpose_y=True, alpha=self.head_dim**-0.5)
+        if attn_mask is not None:
+            # TODO(guosheng): support bool mask
+            # product = product * attn_mask
+            # mask_score = (attn_mask - 1.0) * 10000.0
+            # product = product + mask_score
```
Comment (wawltor) [inline PaddleNLP/paddlenlp/transformers/gpt2/modeling.py]: 这里的注释代码删除掉

File: PaddleNLP/paddlenlp/transformers/gpt2/tokenizer.py [hunk 20]
```diff
@@ -0,0 +1,462 @@
...
+            self.text += [o.token for o in other]
+            self.original_text += [o.token for o in other]
+        elif isinstance(other, Tokenization):
+            self.tokenization.extend(other.tokenization)
+            self.text += other.text
+            self.original_text += other.original_text
+        else:
+            self.tokenization.extend(other)
+        return self
+
```
Comment (wawltor) [inline PaddleNLP/paddlenlp/transformers/gpt2/tokenizer.py]: 这里的Tokenization 看看能不能去掉，太丑了

File: PaddleNLP/paddlenlp/transformers/gpt2/tokenizer.py [hunk 21]
```diff
@@ -0,0 +1,462 @@
...
+    def tokenize(self, text):
+        """ Tokenize a string. """
+        seg_list = [
+            x.translate(self.translator) for x in jieba.cut(text, cut_all=False)
+        ]
+        new_seg = " ".join(seg_list)
+        return self.sp.encode(new_seg)
+
+    def convert_tokens_to_ids(self, text):
```
Comment (wawltor) [inline PaddleNLP/paddlenlp/transformers/gpt2/tokenizer.py:142]: 这块保持和其他的Tokenizer一致吧，把convert_tokens_to_ids改成encode，把convert_ids_to_tokens改成decode

File: PaddleNLP/examples/language_model/gpt2/run_pretrain.py (hunk 15 above)
Comment (ZHUI) [reply PaddleNLP/examples/language_model/gpt2/run_pretrain.py]: done

File: PaddleNLP/examples/language_model/gpt2/README.md (hunk 1 above)
Comment (ZHUI) [reply PaddleNLP/examples/language_model/gpt2/README.md]: 可以给一下具体链接吗？
